### PR TITLE
Introducing backends

### PR DESCRIPTION
**Pull request by kurapov-peter**
This PR introduces a new component - a Backend. The goal of the refactoring is to clean up code generation from any device specific implementation details by providing appropriate methods in a corresponding backend as well as `CodeGenerator` and `Executor` decoupling. Here I start with unifying the native code generation paths for CPU and CUDA.

### CODE REVIEW

File: QueryEngine/TableFunctions/TableFunctionCompilationContext.cpp [hunk 1]
```diff
@@ -527,33 +528,27 @@ std::shared_ptr<CompilationContext> TableFunctionCompilationContext::finalize(
...
-                                                /*is_gpu_smem_used=*/false,
-                                                co,
-                                                gpu_target);
-  } else {
-    auto ee =
-        CodeGenerator::generateNativeCPUCode(entry_point_func_, {entry_point_func_}, co);
-    auto cpu_code = std::make_shared<CpuCompilationContext>(std::move(ee));
-    cpu_code->setFunctionPointer(entry_point_func_);
-    code = cpu_code;
+    target = {executor_->cudaMgr(), executor_->blockSize(), cgen_state, false};
```
Comment (alexbaden) [inline QueryEngine/TableFunctions/TableFunctionCompilationContext.cpp:535]: Why create the empty target on line 531 above?

**Review by alexbaden — APPROVED** [review]
Looks good, just one comment.

File: QueryEngine/Compiler/Backend.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (ienkovich) [inline QueryEngine/Compiler/Backend.cpp:1]: Didn't we want to switch to SPDX ids?
```
/**
 * Copyright 2021 OmniSci, Inc.
 * Copyright (C) 2022 Intel Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */
```

Comment (kurapov-peter) [reply QueryEngine/Compiler/Backend.cpp:1]: I just copied it over from somewhere, is it incorrect?

File: QueryEngine/TableFunctions/TableFunctionCompilationContext.cpp (hunk 1 above)
Comment (kurapov-peter) [reply QueryEngine/TableFunctions/TableFunctionCompilationContext.cpp:535]: An empty target is now used in CPU case for the `compiler::getBackend`.

File: QueryEngine/Compiler/Backend.cpp (hunk 2 above)
Comment (ienkovich) [reply QueryEngine/Compiler/Backend.cpp:1]: Can't say it's incorrect. I just noticed it's different from what I used to see in other files (no blank lines, no '*' in each line). And then I remembered we also wanted to switch to SPDX. Not a critical thing I believe. We can always make a global replacement later.

Comment (kurapov-peter) [reply QueryEngine/Compiler/Backend.cpp:1]: Ok!

**Pull request merged**